### PR TITLE
Global annotations setting for all upstreams

### PR DESCRIPTION
**Pull request by mkazin**
# Description

Adds a new annotations field in Gloo Settings.UpstreamSettings to be applied to every created upstream.

# Context

This new feature allows customers to apply global upstream configuration at a central location rather than repeat it on each individual service.

# Checklist:

- [X] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [X] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [X] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [X] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/6028

### CODE REVIEW

File: projects/gloo/api/v1/settings.proto [hunk 1]
```diff
@@ -450,6 +450,9 @@ message Settings {
 message UpstreamOptions {
     // Default ssl parameter configuration to use for upstreams
     SslParameters ssl_parameters = 1;
+
+    // Annotations to apply to all upstreams
```
Comment (jenshu) [inline projects/gloo/api/v1/settings.proto:454]: we should clarify in the comments here what happens if both sslParameters and globalAnnotations both specify ssl parameter values - which one takes precedence?

File: projects/gloo/pkg/plugins/kubernetes/serviceconverter/use_ssl_annotation_converter.go [hunk 2]
```diff
@@ -42,15 +50,15 @@ func (u *UseSslConverter) ConvertService(svc *kubev1.Service, port kubev1.Servic
 	return nil
 }
 
-func upstreamSslConfigFromService(svc *kubev1.Service, svcPort kubev1.ServicePort) *v1.UpstreamSslConfig {
+func upstreamSslConfigFromAnnotations(annotations map[string]string, svc *kubev1.Service, svcPort kubev1.ServicePort) *v1.UpstreamSslConfig {
```
Comment (jackstine) [inline projects/gloo/pkg/plugins/kubernetes/serviceconverter/use_ssl_annotation_converter.go:53]: nit: a little comment on this function

File: projects/gloo/api/v1/settings.proto (hunk 1 above)
Comment (elcasteel) [reply projects/gloo/api/v1/settings.proto:454]: Also consider adding a list or example of settings that can be configured with global annotations.

Comment (mkazin) [reply projects/gloo/api/v1/settings.proto:454]: After looking at SslParameters, I believe the annotations contain different values so there's no issue:

These are the defined annotations:
```
const GlooSslSecretAnnotation = "gloo.solo.io/sslService.secret"
const GlooSslTlsCertAnnotation = "gloo.solo.io/sslService.tlsCert"
const GlooSslTlsKeyAnnotation = "gloo.solo.io/sslService.tlsKey"
const GlooSslRootCaAnnotation = "gloo.solo.io/sslService.rootCa"
```

SslParameters is defined with these:
```
type SslParameters struct {
...
	MinimumProtocolVersion SslParameters_ProtocolVersion `protobuf:"varint,1,opt,name=minimum_protocol_version,json=minimumProtocolVersion,proto3,enum=gloo.solo.io.SslParameters_ProtocolVersion" json:"minimum_protocol_version,omitempty"`
	MaximumProtocolVersion SslParameters_ProtocolVersion `protobuf:"varint,2,opt,name=maximum_protocol_version,json=maximumProtocolVersion,proto3,enum=gloo.solo.io.SslParameters_ProtocolVersion" json:"maximum_protocol_version,omitempty"`
	CipherSuites           []string                      `protobuf:"bytes,3,rep,name=cipher_suites,json=cipherSuites,proto3" json:"cipher_suites,omitempty"`
	EcdhCurves             []string                      `protobuf:"bytes,4,rep,name=ecdh_curves,json=ecdhCurves,proto3" json:"ecdh_curves,omitempty"`
```

File: projects/gloo/pkg/plugins/kubernetes/uds_convert.go [hunk 3]
```diff
@@ -74,6 +77,19 @@ func (uc *KubeUpstreamConverter) CreateUpstream(ctx context.Context, svc *kubev1
 	return us
 }
 
+// Global Annotations are a set of annotations defined in Gloo's Settings which are applied to all Upstreams
+func applyGlobalAnnotations(ctx context.Context, us *v1.Upstream) {
+	globalAnnotations := settingsutil.MaybeFromContext(ctx).GetUpstreamOptions().GetGlobalAnnotations()
+	upstreamAnnotations := us.GetMetadata().GetAnnotations()
+	if globalAnnotations == nil || upstreamAnnotations == nil {
```
[the remark this answers is not in the document]
Comment (mkazin) [reply projects/gloo/pkg/plugins/kubernetes/uds_convert.go]: This code doesn't exist following the rework.

File: projects/gloo/pkg/plugins/kubernetes/serviceconverter/use_ssl_annotation_converter.go (hunk 2 above)
Comment (mkazin) [reply projects/gloo/pkg/plugins/kubernetes/serviceconverter/use_ssl_annotation_converter.go]: This change has been reverted

**Comment by elcasteel** [comment]
Looks good to me, but I will give Gunnar a chance to review as well.

**Comment by gunnar-solo** [comment]
/kick [Step #10 - "test": [Fail] CSRF defined on listener only on listener [It] should succeed with allowed origin](https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=8bf9f308-1552-4fdd-b465-4a1030bdc1b7)

**Pull request merged**